internal/xpkg/v2/build.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package xpkg
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"strings"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"gopkg.in/yaml.v2"
+	crd "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/parser"
+
+	pkgmetav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	"github.com/crossplane/crossplane/apis/pkg/meta/v1beta1"
+	xpkgv1 "github.com/crossplane/crossplane/internal/xpkg"
+	"github.com/crossplane/crossplane/internal/xpkg/v2/parser/examples"
+)
+
+const (
+	errParserPackage     = "failed to parse package"
+	errParserExample     = "failed to parse examples"
+	errLintPackage       = "failed to lint package"
+	errInitBackend       = "failed to initialize package parsing backend"
+	errTarFromStream     = "failed to build tarball from stream"
+	errLayerFromTar      = "failed to convert tarball to image layer"
+	errDigestInvalid     = "failed to get digest from image layer"
+	errBuildImage        = "failed to build image from layers"
+	errConfigFile        = "failed to get config file from image"
+	errMutateConfig      = "failed to mutate config for image"
+	errBuildObjectScheme = "failed to build scheme for package encoder"
+	errParseAuth         = "an auth extension was supplied but could not be parsed"
+	errAuthNotAnnotated  = "an auth extension was supplied but but the " + providerConfigKind + " object could not be found"
+	errNotExactlyOneMeta = "package must contain exactly one meta object"
+	authMetaAnno         = "auth.upbound.io/group"
```

Comment:
is this something we need here ?

Reply:
given this is used to configure `ProviderConfig` functionality in `console.upbound.io`, probably not. It does imply that any provider in the future that wants to take advantage of that functionality would have to be built with [up](https://github.com/upbound/up), which is the case today.

Reply:
Removed the auth extension option then. We can leave this to be `up` specific